### DISCUSSION TITLE
Question regarding buttons disabling

### DISCUSSION THREAD

**Discussion by BareGeode** [AvaloniaUI/Avalonia#16552] [Q&A] [answered]
Hi. I'm a complete beginner to Avalonia, so I'm really confused by one of the example projects which I found in the documentation called [FileOps](https://github.com/AvaloniaUI/AvaloniaUI.QuickGuides/tree/main/FileOps). It's a simple text file editor where there's a button to open the file viewer to select a text file. When you use this button and the file viewer pops up, the button disables. Why does this happen? I can't find any code in the project itself which disables the button, so I guess it must be something related to how the command works, but I can't find any info about it online.

**Comment by BAndysc** [comment, marked as answer]
This sample uses `CommunityToolkit.Mvvm` with [source generator](https://learn.microsoft.com/pl-pl/dotnet/communitytoolkit/mvvm/relaycommand) attribute `[RelayCommand]` on a method. This source generator creates a property `[MethodName]Command`, which depending on the method will be `CommunityToolkit.Mvvm.Input.AsyncRelayCommand` or `CommunityToolkit.Mvvm.Input.RelayCommand`.

As you can guess, for Task methods, it generates `AsyncRelayCommand`. And that's how `AsyncRelayCommand` works in `CommunityToolkit.Mvvm` by default - while the task is being executed, `CanExecute` returns false.

You can disable this behaviour via attribute parameter: `[RelayCommand(AllowConcurrentExecutions = true)]`

You can see the `AsyncRelayCommand` class [implementation here](https://github.com/CommunityToolkit/dotnet/blob/main/src/CommunityToolkit.Mvvm/Input/AsyncRelayCommand.cs#L276). Please note this is not related to Avalonia at all. Avalonia doesn't care if this is an async method or not - all logic here belongs to `CommunityToolkit.Mvvm`. Avalonia only monitors `CanExecute` method, that's all.
